client/state/signup/progress/reducer.js

```diff
@@ -3,132 +3,120 @@
 	SIGNUP_PROGRESS_SAVE_STEP,
 	SIGNUP_PROGRESS_SUBMIT_STEP,
 } from 'state/action-types';
 import { createReducer } from 'state/utils';
 import { schema } from './schema';
-import userFactory from 'lib/user';
 
 const debug = debugFactory( 'calypso:state:signup:progress:reducer' );
 
 //
-// Action handlers
+// Helper Functions
 //
-function overwriteSteps( state, { steps = [] } ) {
-	// When called without action.steps, this is basically a state reset function.
-	return Array.isArray( steps ) ? steps : [];
-}
+const addStep = ( state, step ) => {
+	debug( `Adding step ${ step.stepName }` );
 
-function completeStep( state, { step } ) {
-	return updateStep( state, { ...step, status: 'completed' } );
-}
+	return {
+		...state,
+		[ step.stepName ]: step,
+	};
+};
+
+const updateStep = ( state, newStepState ) => {
+	const { stepName, status } = newStepState;
+	const stepState = get( state, stepName );
 
-function invalidateStep( state, { step, errors } ) {
-	const newStep = { ...step, errors, status: 'invalid' };
-	if ( find( state, { stepName: newStep.stepName } ) ) {
-		return updateStep( state, newStep );
+	if ( ! stepState ) {
+		return state;
 	}
-	return addStep( state, newStep );
-}
 
-function processStep( state, { step } ) {
-	return updateStep( state, { ...step, status: 'processing' } );
-}
+	debug( `Updating step ${ stepName }` );
+
+	if ( status === 'pending' || status === 'completed' ) {
+		// This can only happen when submitting a step
+		//
+		// Steps that are resubmitted may decide to omit the `wasSkipped` status of a step if e.g.
+		// the user goes back and chooses to not skip a step. If a step is submitted without it,
+		// we explicitly remove it from the step data.
+		const { wasSkipped, ...commonStepArgs } = stepState;
+
+		return {
+			...state,
+			[ stepName ]: {
+				...commonStepArgs,
+				...newStepState,
+			},
+		};
+	}
 
-function removeUnneededSteps( state, { flowName } ) {
-	let flowSteps = [];
-	const user = userFactory();
+	return {
+		...state,
+		[ stepName ]: {
+			...stepState,
+			...newStepState,
+		},
+	};
+};
 
-	flowSteps = get( flows, [ flowName, 'steps' ], [] );
+//
+// Action handlers
+//
 
-	if ( user && user.get() ) {
-		flowSteps = flowSteps.filter( item => item !== 'user' );
-	}
+// When called without action.steps, this is basically a state reset function.
+const overwriteSteps = ( state, { steps = {} } ) => keyBy( steps, 'stepName' );
 
-	return state.filter( step => flowSteps.includes( step.stepName ) );
-}
+const completeStep = ( state, { step } ) => updateStep( state, { ...step, status: 'completed' } );
 
-function saveStep( state, { step } ) {
-	if ( find( state, { stepName: step.stepName } ) ) {
-		return updateStep( state, step );
-	}
+const invalidateStep = ( state, { step, errors } ) => {
+	const newStepState = { ...step, errors, status: 'invalid' };
 
-	return addStep( state, { ...step, status: 'in-progress' } );
-}
+	return has( state, step.stepName )
```

Comment:
Nitpick: Using lodash for simple stuff like that adds an extra step of "better look it up in lodash docs" while reviewing. `step.stepName in state` accomplishes the same thing and is easier to read

Reply:
I'll leave this one be, this suggestion would error in the case where `state` is `undefined`.
I know thats hugely unlikely, but I always prefer to be consistent with patterns like this.
Also, I'd argue that any dev has to look up documentation when touching new code and we lodash is well used throughout Calypso, so the learning opportunity wouldn't hurt and could encourage others to use safer methods.